### PR TITLE
Adds tests and potential fixes for #10771 and #11143

### PR DESCRIPTION
**Pull request by akitchen**
Unfortunately a brief spin of the run loop is required for some of these tests; we should take a critical look at which ones are worth keeping around once we're satisfied with the fixes they are helping to drive, in order to avoid slowing down our test suite too much.

/ref #10771 #11143
/cc @asheemmamoowala @julianrex @1ec5 @lilykaiser

### CODE REVIEW

**Comment by julianrex** [comment]
@asheemmamoowala @julianrex I have a potential fix for these errors (and integration tests not running) at https://github.com/mapbox/mapbox-gl-native/compare/gl-layer-memory-fix...jrex-gl-layer-memory-fix.

All the tests run, and I played with the iOS app for a while and all seemed good, but obviously this could have some significant impact, and there may be other cases that I haven't considered.

Can you take a look? If it looks good, we could merge into master, and then cherry-pick into boba?

To get the tests to run on device, I had to add Mapbox.framework to the test app as an embedded framework (rather than adding it to the integration framework).

**Comment by akitchen** [comment]
@jmkiley do you have a working example which we could use to verify these fixes? I thought maybe you did, from your previous investigations...

File: platform/darwin/src/MGLStyleLayer.mm [hunk 1]
```diff
@@ -50,6 +54,10 @@ - (void)removeFromStyle:(MGLStyle *)style
 {
     if (self.rawLayer == style.rawStyle->getLayer(self.identifier.UTF8String)) {
         _pendingLayer = style.rawStyle->removeLayer(self.identifier.UTF8String);
+
+        // Pair the retain above, and release self, since we're now removed from the collection
+        CFTypeRef toRelease = (__bridge CFTypeRef)self;
+        CFBridgingRelease(toRelease);
```
[the remark this answers is not in the document]
Comment (julianrex) [reply platform/darwin/src/MGLStyleLayer.mm]: Scratch that - I think we need some more clean-up.

**Comment by jmkiley** [comment]
I used this branch with [my repro branch](https://github.com/mapbox/mapbox-gl-native/tree/jmkiley-debug-opengllayer) and it seemed to be 👍

File: platform/ios/Integration Tests/MBGLIntegrationTests.m [hunk 2]
```diff
@@ -53,25 +54,161 @@ - (MGLStyle *)style {
...
+            
+            retrievedLayer = [self.style layerWithIdentifier:@"gl-layer"];
+            XCTAssertNotNil(retrievedLayer);
+            [self.style removeLayer:retrievedLayer];
+
+            // We need to run the runloop for a little while so that the following assert will be correct
+            // this is because although the layer has been removed from the style, there's still a pending
+            // render (deinitialize) call, that will needs to be handled, which will finally release the
+            // layer (and then the layer will be dealloc'd when the autorelease pool drains)
+            [[NSRunLoop currentRunLoop] runUntilDate:[[NSDate date] dateByAddingTimeInterval:waitInterval]];
```
Comment (1ec5) [inline platform/ios/Integration Tests/MBGLIntegrationTests.m]: If we expect the layer to be released as of the next render, how about putting in an expectation that gets satisfied upon the next call to `-[MGLMapViewDelegate mapViewDidFinishRenderingFrame:fullyRendered:]`?

**Comment by asheemmamoowala** [comment]
Incorporated into: #11553

**Comment by akitchen** [comment]
I believe this can be closed now that #11553 has landed... please reopen if I'm mistaken.

**Pull request closed without merging**